### PR TITLE
Make traversing derived relationships work when using `Bytes` ids

### PR DESCRIPTION
**Pull request by lutter**
There will also be tests that expose this issue, and make sure it stays fixed. But since that involves an extensive rewrite of `graphql/tests/query.rs`, it will happen in a separate PR so we can at least roll out this fix first.

### CODE REVIEW

File: store/postgres/src/relational_queries.rs [hunk 1]
```diff
@@ -2171,6 +2203,26 @@ impl<'a> FilterCollection<'a> {
...
+            FilterCollection::SingleWindow(window) => Ok(Some(window.parent_type())),
+            FilterCollection::MultiWindow(windows, _) => {
+                let parent_types: HashSet<_> =
+                    HashSet::from_iter(windows.into_iter().map(|window| window.parent_type()));
+                if parent_types.len() > 1 {
+                    return Err(graph::constraint_violation!(
+                        "all implementors of an interface must use the same type for their `id`"
+                    ));
+                }
+                Ok(parent_types.iter().next().map(|t| t.to_owned()))
```
Comment (neysofu) [inline store/postgres/src/relational_queries.rs:2222]: We could also use `all_equal()`, as a possibly clearer alternative to the `HashSet` and `> 1` logic:

```suggestion
                if windows.iter().map(FilterWindow::parent_type).all_equal() {
                    Ok(Some(windows[0].parent_type().to_owned()))
                } else {
                    Err(graph::constraint_violation!(
                        "all implementors of an interface must use the same type for their `id`"
                    ))
                }
```

Just food for thought; feel free to ignore this.

File: store/postgres/src/relational_queries.rs [hunk 2]
```diff
@@ -473,8 +474,32 @@ impl EntityData {
...
+                                    out.insert_entity_data("g$parent_id".to_owned(), value);
+                                } else {
+                                    // A query that does not have parents
+                                    // somehow returned parent ids. We have no
+                                    // idea how to deserialize that
+                                    if parent_type.is_none() {
+                                        return Err(graph::constraint_violation!(
+                                            "query unexpectedly produces parent ids"
+                                        ));
+                                    }
```
Comment (neysofu) [inline store/postgres/src/relational_queries.rs:495]: Why do we need this check? I think it's redundant, at this point in the code we already know that `parent_type.is_none()`.

**Review by neysofu — CHANGES_REQUESTED** [review]
Looks good, just some minor comments.

File: graph/src/env/store.rs [hunk 3]
```diff
@@ -184,4 +190,6 @@ pub struct InnerStore {
     connection_idle_timeout_in_secs: u64,
     #[envconfig(from = "GRAPH_STORE_WRITE_QUEUE", default = "5")]
     write_queue_size: usize,
+    #[envconfig(from = "GRAPH_DISABLE_ERROR_FOR_TOPLEVEL_PARENTS", default = "false")]
+    disable_error_for_toplevel_parents: bool,
```
[the remark this answers is not in the document]
Comment (lutter) [reply graph/src/env/store.rs:194]: Nice! I didn't know that existed. Changed.

File: store/postgres/src/relational_queries.rs (hunk 2 above)
Comment (lutter) [reply store/postgres/src/relational_queries.rs:495]: Yes, that's absolutely right. Fixed.

File: store/postgres/src/relational_queries.rs (hunk 1 above)
Comment (lutter) [reply store/postgres/src/relational_queries.rs:2222]: I didn't know about `all_equal`. I like your way a lot better. Changed.

**Comment by lutter** [comment]
Addressed all review comments.

**Review by neysofu — APPROVED** [review]
Looks great! Feel free to merge 🌈

**Pull request merged**